yacl/crypto/primitives/tpre/capsule.cc

```diff
@@ -0,0 +1,214 @@
+// Copyright 2023 Chengfang Financial Technology Co., Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "yacl/crypto/primitives/tpre/capsule.h"
+
+#include "yacl/base/exception.h"
+#include "yacl/crypto/base/hash/hash_utils.h"
+#include "yacl/crypto/primitives/tpre/hash.h"
+#include "yacl/crypto/primitives/tpre/kdf.h"
+#include "yacl/utils/scope_guard.h"
+
+namespace yacl::crypto {
+
+// Encapsulate(pkA)->(K,capsule)
+std::pair<Capsule::CapsuleStruct, std::vector<uint8_t>> Capsule::EnCapsulate(
+    const std::unique_ptr<EcGroup>& ecc_group,
+    const Keys::PublicKey& delegating_public_key) const {
+  MPInt zero_bn(0);
+  MPInt order = ecc_group->GetOrder();
+  MPInt r;
+  MPInt::RandomLtN(order, &r);
+  MPInt u;
+  MPInt::RandomLtN(order, &u);
+
+  EcPoint E = ecc_group->MulBase(r);
+  EcPoint V = ecc_group->MulBase(u);
+  std::string E_string_join_V_sting = ecc_group->GetAffinePoint(E).ToString() +
```

Comment:
`ecc_group->GetAffinePoint(E)` is so heavy and better for debug info. Suggest `std::string(ecc_group->SerializePoint(E))` instead of `ecc_group->GetAffinePoint(E).ToString()`. **Same for all files**